### PR TITLE
gh-98783: Fix crashes when `str` subclasses where used in `_PyUnicode_Equal`

### PR DESCRIPTION
**Pull request by sobolevn**
Thanks a lot for your help, @sweeneyde!

* Issue: gh-98783

### CODE REVIEW

**Comment by sobolevn** [comment]
Windows failure is unrelated, but I will retrigger anyway.

**Review by sweeneyde — APPROVED** [review]

File: Misc/NEWS.d/next/Core and Builtins/2022-10-28-14-52-55.gh-issue-98783.iG0kMs.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Fix multiple crashes when ``str`` subclasses where used instead of ``str``
```
Comment (sweeneyde) [inline Misc/NEWS.d/next/Core and Builtins/2022-10-28-14-52-55.gh-issue-98783.iG0kMs.rst]: Maybe mention that these were assertion failures in debug builds only?

Comment (sobolevn) [reply Misc/NEWS.d/next/Core and Builtins/2022-10-28-14-52-55.gh-issue-98783.iG0kMs.rst]: Done!

File: Misc/NEWS.d/next/Core and Builtins/2022-10-28-14-52-55.gh-issue-98783.iG0kMs.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Fix multiple crashes in debug mode when ``str`` subclasses
+where used instead of ``str`` itself.
```
Comment (sweeneyde) [inline Misc/NEWS.d/next/Core and Builtins/2022-10-28-14-52-55.gh-issue-98783.iG0kMs.rst]:
```suggestion
are used instead of ``str`` itself.
```

**Comment by sobolevn** [comment]
Thank you for your help!

**Pull request merged**